File: maheshwari_motors_app/lib/app/data/models/bill_model.dart
class BillModel {
  final String id;
  final String billNo;
  final DateTime date;
  final String? partyId;
  final String? partyName;
  final double amount;
  final double paidAmount;
  final double returnAmount;
  final String paymentStatus;
  final List<String> challanIds;
  final bool skipStockCalculation;
  final int isGst;
  final double? balance;

  BillModel({
    required this.id,
    required this.billNo,
    required this.date,
    this.partyId,
    this.partyName,
    required this.amount,
    this.paidAmount = 0,
    this.returnAmount = 0,
    this.paymentStatus = 'due',
    this.challanIds = const [],
    this.skipStockCalculation = false,
    this.isGst = 1,
    this.balance,
  });

  factory BillModel.fromJson(Map<String, dynamic> json) {
    final partyData = json['party_id'];
    return BillModel(
      id: json['_id'] ?? '',
      billNo: json['bill_no'] ?? '',
      date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
      partyId: partyData is Map ? partyData['_id'] : partyData?.toString(),
      partyName: partyData is Map ? partyData['name'] : null,
      amount: (json['amount'] ?? 0).toDouble(),
      paidAmount: (json['paid_amount'] ?? 0).toDouble(),
      returnAmount: (json['return_amount'] ?? 0).toDouble(),
      paymentStatus: json['payment_status'] ?? 'due',
      challanIds: json['challan_ids'] != null
          ? List<String>.from(
              (json['challan_ids'] as List).map(
                (e) => e is String ? e : e['_id'] ?? '',
              ),
            )
          : [],
      skipStockCalculation: json['skip_stock_calculation'] ?? false,
      isGst: json['is_gst'] ?? 1,
      balance: json['balance']?.toDouble(),
    );
  }

  double get balanceAmount => balance ?? (amount - paidAmount);

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      '_id': id,
      'bill_no': billNo,
      'date': date.toIso8601String(),
      'amount': amount,
      'paid_amount': paidAmount,
      'return_amount': returnAmount,
      'payment_status': paymentStatus,
      'challan_ids': challanIds,
      'skip_stock_calculation': skipStockCalculation,
      'is_gst': isGst,
    };
    if (partyId != null) map['party_id'] = partyId;
    if (balance != null) map['balance'] = balance;
    return map;
  }
}

File: backend/src/models/bill.model.js
import mongoose from "mongoose";

/**
 * Bill Model (Updated)
 * - Added skip_stock_calculation: when true, this bill doesn't affect physical stock
 * - Use case: Selling from GST stock via NON_GST firm (virtual stock transfer)
 */
const billSchema = new mongoose.Schema(
  {
    bill_no: { type: String, required: true },
    date: { type: Date, required: true, default: Date.now },
    party_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Party",
      required: true,
    },
    amount: { type: Number, required: true },
    paid_amount: { type: Number, default: 0 },
    return_amount: { type: Number, default: 0 },
    payment_status: {
      type: String,
      enum: ["due", "paid", "overpaid"],
      default: "due",
    },
    challan_ids: [{ type: mongoose.Schema.Types.ObjectId, ref: "Challan" }],

    // When true, this bill does NOT deduct from physical stock
    // Use case: Virtual stock transfer between GST/NON_GST firms
    skip_stock_calculation: { type: Boolean, default: false },

    // GST classification: 1 = GST bill, 0 = non-GST bill
    is_gst: { type: Number, enum: [0, 1] },

    // DEPRECATED: user_id - keeping for backward compat during migration
    user_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
    },
    firm_id: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Firm",
      required: true,
    },
  },
  { timestamps: true, id: false },
);

billSchema.virtual("balance").get(function () {
  return this.amount - this.paid_amount;
});

billSchema.set("toJSON", { virtuals: true });
billSchema.set("toObject", { virtuals: true });

billSchema.index({ bill_no: 1, firm_id: 1 }, { unique: true });

export default mongoose.model("Bill", billSchema);
